File: swift/Sources/MudbaseShowcaseEvents/Models/ActivityEntry.swift
import Foundation

/// Mirrors `web/src/types/activity.ts`'s `ActivityAction` union and `ACTIVITY_LABELS` map exactly.
enum ActivityAction: String, Sendable, Equatable, CaseIterable {
    case bookingConfirmed = "booking_confirmed"
    case bookingWaitlisted = "booking_waitlisted"
    case bookingCancelled = "booking_cancelled"
    case bookingPromoted = "booking_promoted"
    case checkedIn = "checked_in"
    case eventCreated = "event_created"
    case eventUpdated = "event_updated"

    var label: String {
        switch self {
        case .bookingConfirmed: return "booked (confirmed)"
        case .bookingWaitlisted: return "joined the waitlist"
        case .bookingCancelled: return "cancelled their booking"
        case .bookingPromoted: return "was promoted from the waitlist"
        case .checkedIn: return "checked in"
        case .eventCreated: return "created this event"
        case .eventUpdated: return "updated this event"
        }
    }
}

/// Mirrors `web/src/types/activity.ts`'s `ActivityDoc` interface exactly.
struct ActivityEntry: Sendable, Equatable, Identifiable {
    let id: String
    let createdAt: Date?
    let eventId: String
    let actorId: String
    let actorName: String
    let action: ActivityAction

    /// Fails when `action` doesn't match a known case — a forward-compatible guard against a
    /// future action this build doesn't know about yet, rather than rendering a blank/garbled row.
    init?(document: MudbaseDocument) {
        guard let action = document.string("action").flatMap(ActivityAction.init(rawValue:)) else { return nil }
        id = document.id
        createdAt = document.createdAt
        eventId = document.string("eventId") ?? ""
        actorId = document.string("actorId") ?? ""
        actorName = document.string("actorName") ?? ""
        self.action = action
    }
}

File: java/src/main/java/dev/mudbase/showcase/events/domain/ActivityEntry.java
package dev.mudbase.showcase.events.domain;

import dev.mudbase.showcase.events.mudbase.DocumentMapper;
import dev.mudbase.showcase.events.support.Formatting;
import java.util.Map;

/**
 * Mirrors the `activity` collection - see plan/build-plan.md. {@link #getDescription()} mirrors
 * the reference web app's {@code ACTIVITY_LABELS} map (../web/src/types/activity.ts) sentence-for-
 * sentence.
 */
public class ActivityEntry {

  private final String id;
  private final String eventId;
  private final String actorId;
  private final String actorName;
  private final String action;
  private final String createdAt;

  private ActivityEntry(
      String id, String eventId, String actorId, String actorName, String action, String createdAt) {
    this.id = id;
    this.eventId = eventId;
    this.actorId = actorId;
    this.actorName = actorName;
    this.action = action;
    this.createdAt = createdAt;
  }

  public static ActivityEntry fromDocument(Map<String, Object> doc) {
    return new ActivityEntry(
        DocumentMapper.getId(doc),
        DocumentMapper.getString(doc, "eventId"),
        DocumentMapper.getString(doc, "actorId"),
        DocumentMapper.getString(doc, "actorName", "Someone"),
        DocumentMapper.getString(doc, "action", ""),
        DocumentMapper.getString(doc, "createdAt"));
  }

  public String getId() {
    return id;
  }

  public String getEventId() {
    return eventId;
  }

  public String getActorId() {
    return actorId;
  }

  public String getActorName() {
    return actorName;
  }

  public String getAction() {
    return action;
  }

  public String getCreatedAt() {
    return createdAt;
  }

  public String getFormattedCreatedAt() {
    return Formatting.formatRelativeTime(createdAt);
  }

  /** Mirrors ../web/src/types/activity.ts `ACTIVITY_LABELS` exactly. */
  public String getDescription() {
    return switch (action) {
      case "booking_confirmed" -> "booked (confirmed)";
      case "booking_waitlisted" -> "joined the waitlist";
      case "booking_cancelled" -> "cancelled their booking";
      case "booking_promoted" -> "was promoted from the waitlist";
      case "checked_in" -> "checked in";
      case "event_created" -> "created this event";
      case "event_updated" -> "updated this event";
      default -> action;
    };
  }
}
